### PR TITLE
Add support for logging with multiple tags

### PR DESCRIPTION
**Pull request by kahgoh**
### Description

Linked issue: #201

This change allows a single `TaggedLogger` to log to multiple tags, instead of having to create individual logger for each tag.

### Definition of Done

- [x] I read [contributing.md](./contributing.md)
- [x] There are no TODOs left in the code
- [x] Code style follows the tinylog standard
- [x] All classes and methods have Javadoc
- [x] Changes are covered by JUnit tests including edge cases, errors, and exception handling
- [x] Maven build works including compiling, tests, and checks (`mvn verify`)
- [x] Changes are committed by a verified email address that is assigned to the GitHub account (https://github.com/settings/emails)

### Documentation

Additions or amendments for the [public documentation](https://github.com/pmwmedia/tinylog/wiki/Documentation):

- [Tags](https://github.com/tinylog-org/tinylog/wiki/Logging#tags) will need to be updated with the addition of the new `tags` method. 

### Agreements

- [x] I agree that my changes will be published under the terms of the [Apache License 2.0](https://github.com/pmwmedia/tinylog/blob/v2.0/license.txt) (mandatory)
- [x] I agree that my GitHub user name will be published in the release notes (optional)

### CODE REVIEW

**Comment by pmwmedia** [comment]
What is the current status? When can I do the review?

**Comment by kahgoh** [comment]
Sorry, its taking me a while. I'm still working updating the Scala API side.

**Comment by pmwmedia** [comment]
@kahgoh Thanks for the status update :)

**Comment by kahgoh** [comment]
I've pushed in the Scala API changes and think its ready for review now. I'll come back to add something for the documentation later when I get the chance.

**Comment by pmwmedia** [comment]
Thank you very much for your commit :) I'm going to review it. Could you fix one compile error in the mean time?

```
/home/runner/work/tinylog/tinylog/tinylog-api/src/main/java/org/tinylog/Logger.java:[92,47] diamond operator is not supported in -source 1.6
```

For Android support, the source code still needs to be Java 6 compatible.

File: tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/Logger.kt [hunk 1]
```diff
@@ -39,27 +39,23 @@ object Logger {
...
 	 * @param tag
 	 * Tag for logger or `null` for receiving an untagged logger
+	 * @param otherTags
+	 * Additional tags for the logger
 	 * @return Logger instance
 	 */
-	fun tag(tag: String?): TaggedLogger {
-		if (tag == null || tag.isEmpty()) {
-			return instance
+	fun tag(tag: String?, vararg otherTags: String?): TaggedLogger {
```
Comment (pmwmedia) [inline tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/Logger.kt]: I would prefer two different methods: `tag(tag: String?)` and `tags(vararg tags: String?)`

File: tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/TaggedLogger.kt [hunk 2]
```diff
@@ -21,23 +21,25 @@ import org.tinylog.provider.ProviderRegistry
...
-	private val minimumLevelCoversTrace = isCoveredByMinimumLevel(tag, Level.TRACE)
-	private val minimumLevelCoversDebug = isCoveredByMinimumLevel(tag, Level.DEBUG)
-	private val minimumLevelCoversInfo  = isCoveredByMinimumLevel(tag, Level.INFO)
-	private val minimumLevelCoversWarn  = isCoveredByMinimumLevel(tag, Level.WARN)
-	private val minimumLevelCoversError = isCoveredByMinimumLevel(tag, Level.ERROR)
+	private val traceTags = getCoveredTags(tags, Level.TRACE)
+	private val debugTags = getCoveredTags(tags, Level.DEBUG)
+	private val infoTags  = getCoveredTags(tags, Level.INFO)
+	private val warnTags  = getCoveredTags(tags, Level.WARN)
+	private val errorTags = getCoveredTags(tags, Level.ERROR)
```
Comment (pmwmedia) [inline tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/TaggedLogger.kt:41]: Actually we need booleans here. By using booleans, the JVM can detect disabled severity levels and remove these calls completely at runtime. This is the primary reason why tinylog is such fas in discarding log entries (see https://tinylog.org/v2/benchmark/#discarding). However, such optimization by the JVM won't work with sets.

Therefore, I propose to calculate the lowest severity level of all tags and store the result as boolean.

File: tinylog-api-scala/src/main/scala/org/tinylog/scala/Logger.scala [hunk 3]
```diff
@@ -22,24 +22,27 @@ import scala.language.experimental.macros
...
 		*
 		* @param tag
 		* Tag for logger or `null` for receiving an untagged logger
+		* @param otherTags
+		* Additional tags for the logger
 		* @return Logger instance
 		*/
-	def tag(tag: String): TaggedLogger = {
-		if (tag == null || tag.isEmpty()) {
+	def tag(tag: String, otherTags: String*): TaggedLogger = {
```
Comment (pmwmedia) [inline tinylog-api-scala/src/main/scala/org/tinylog/scala/Logger.scala]: I would prefer two different methods: `tag(tag: String)` and `tags(tags: String*)`

File: tinylog-api/src/main/java/org/tinylog/Logger.java [hunk 4]
```diff
@@ -56,23 +59,44 @@ private Logger() {
 	 *
 	 * @param tag
 	 *            Tag for logger or {@code null} for receiving an untagged logger
+	 * @param otherTags
+	 *            Additional tags for the logger
 	 * @return Logger instance
 	 */
-	public static TaggedLogger tag(final String tag) {
-		if (tag == null || tag.isEmpty()) {
+	public static TaggedLogger tag(final String tag, final String...otherTags) {
```
Comment (pmwmedia) [inline tinylog-api/src/main/java/org/tinylog/Logger.java]: I would prefer two different methods: `tag(final String tag)` and `tags(final String... tags)`

File: tinylog-api/src/main/java/org/tinylog/TaggedLogger.java [hunk 5]
```diff
@@ -35,27 +39,34 @@
...
-	private final boolean minimumLevelCoversTrace;
-	private final boolean minimumLevelCoversDebug;
-	private final boolean minimumLevelCoversInfo;
-	private final boolean minimumLevelCoversWarn;
-	private final boolean minimumLevelCoversError;
+	private final Set<String> traceTags;
+	private final Set<String> debugTags;
+	private final Set<String> infoTags;
+	private final Set<String> warnTags;
+	private final Set<String> errorTags;
```
Comment (pmwmedia) [inline tinylog-api/src/main/java/org/tinylog/TaggedLogger.java:48]: We need booleans here (see my comment in the Kotlin implementation).

File: tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/TaggedLogger.kt [hunk 6]
```diff
@@ -56,9 +58,7 @@ class TaggedLogger internal constructor(private val tag: String?) {
 	 * Any object with a meaningful [Any.toString] method
 	 */
 	fun trace(message: Any?) {
-		if (minimumLevelCoversTrace) {
-			provider.log(stackTraceDepth, tag, Level.TRACE, null, null, message)
-		}
+		traceTags.forEach { provider.log(stackTraceDepth, it, Level.TRACE, null, null, message) }
```
Comment (pmwmedia) [inline tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/TaggedLogger.kt]: Does the the lambda call add stack trace element!? I'm not sure. Maybe we need to add `+1` to `stackTraceDepth`!?

File: tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/TaggedLogger.kt [hunk 7]
```diff
@@ -790,11 +790,33 @@ class TaggedLogger internal constructor(private val tag: String?) {
...
 
+	/**
+	 * Checks if any of the tags in a given {@link Set} is enabled for a given {@link Level}.
+	 *
+	 * @param tags  {@link Set} of tags to check
+	 * @param level the log level that at least one of the tags must be enabled
+	 * @return {@code true} if the level is enabled for at least one of the tags in the given {@link Set}. otherwise, {@code false}
+	 */
+	private fun anyEnabled(tags: Set<String?>, level: Level): Boolean {
+		return tags.any{provider.isEnabled(stackTraceDepth, it, level)}
```
Comment (pmwmedia) [inline tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/TaggedLogger.kt]: The method `anyEnabled()` adds a stack trace element to the stack trace, so `stackTraceDepth + 1` is required.

File: tinylog-api-kotlin/src/test/kotlin/org/tinylog/kotlin/LoggerTest.kt [hunk 8]
```diff
@@ -1005,7 +996,33 @@ class LoggerTest {
...
+				Assertions.fail("tags (${tags}) is expected to be a collection, but isn't")
+			}
+		}
+
+		/**
+		 * Verifies that [Logger.tag] returns the same tagged instance of [TaggedLogger] for each
+		 * set of tags with more than one tag.
+		 */
+		@Test
+		fun taggedMultiple() {
```
Comment (pmwmedia) [inline tinylog-api-kotlin/src/test/kotlin/org/tinylog/kotlin/LoggerTest.kt:1019]: Good test! Could you also add an additional test for mixing tags with `null`?

**Review by pmwmedia — CHANGES_REQUESTED** [review]

**Comment by pmwmedia** [comment]
I have just finished my review. Your code looks very well! I have found just a few minors to change.

File: tinylog-api/src/main/java/org/tinylog/Logger.java (hunk 4 above)
Comment (kahgoh) [reply tinylog-api/src/main/java/org/tinylog/Logger.java]: I was thinking if we went with the two methods, it would be possible to call the method with no tags (e.g. `Logger.tag().info(...)`) - something that wouldn't be possible with the single method. 

If we go with the two different methods, should we return the one from the `instance` field if its called with no tags? The other possibility I could think of is return a `TaggedLogger` with no tags, but I'm not sure how meaningful that is.

Comment (pmwmedia) [reply tinylog-api/src/main/java/org/tinylog/Logger.java]: Actually, `TaggedLogger` without tags is already in use by some users. In some use cases, an instance of tinylog's static `Logger` class is required. In this case, a `TaggedLogger` instance without any tags is the way to go.

I propose to see `Logger.tags()` as an alias for `Logger.tag(null)`. For me, it would be even legal to call `Logger.tags(null, "foo", "bar")`. In this case, the log entry will be logged as untagged log entry as well as tagged by `foo` and `bar`.

Comment (kahgoh) [reply tinylog-api/src/main/java/org/tinylog/Logger.java]: That sounds good to me :)

File: tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/TaggedLogger.kt (hunk 6 above)
Comment (kahgoh) [reply tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/TaggedLogger.kt]: The call doesn't add a stack trace element, but I am thinking it might be better to refactor to use an explicit for-loop anyway.

The `forEach` call doesn't add a stack trace element because it is an [inline function](https://kotlinlang.org/docs/inline-functions.html#noinline) that inlines a for-loop (see [Kotlin documentation](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/for-each.html)). The compiler will inline a for-loop in. So, it shouldn't add a stack trace element in.

To further test this, I quickly tested this with the following code:
```kotlin
fun main(args: Array<String>) {
    doLog(1, 2, 3)
}

fun doLog(vararg values: Int) {
    values.toSet().forEach { v ->
        println(v)
        if (v == 3) {
            NullPointerException("Got 3").printStackTrace()
        }
    }
}
```

The output from this was:
```
1
2
3
java.lang.NullPointerException: Got 3
	at test.tinylog.LoopTesterKt.doLog(LoopTester.kt:13)
	at test.tinylog.LoopTesterKt.main(LoopTester.kt:6)
```

Notice the `forEach` didn't appear in the stack trace!

Anyway, this discussion was making me think about refactoring to using an explicit for-loop anyway because I think it will be easier to think about and it won't rely on `forEach` being an inline function.

**Comment by kahgoh** [comment]
Thanks for the feedback and review! I have just finished and pushed the changes.

File: tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/TaggedLogger.kt [hunk 9]
```diff
@@ -19,25 +19,34 @@ import org.tinylog.format.AdvancedMessageFormatter
...
- * @param tag
+ * @param tags
  * Case-sensitive tag for logger instance
  * @see Logger.tag
+ * @see Logger.tags
  */
-class TaggedLogger internal constructor(private val tag: String?) {
+class TaggedLogger internal constructor(private val tags: Set<String?>) {
+
+	constructor(tag: String?) : this(setOf(tag))
```
Comment (pmwmedia) [inline tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/TaggedLogger.kt]: Do we really still need this constructor? If yes, could you declare the constructor as internal? I want to prohibit that users create their own instance.

**Review by pmwmedia — CHANGES_REQUESTED** [review]

File: tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/TaggedLogger.kt (hunk 6 above)
Comment (pmwmedia) [reply tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/TaggedLogger.kt]: Thank you very much for the investigation and the great explanation. Great to know that it is an inline function :) I agree with you that we should not rely on it.

**Comment by pmwmedia** [comment]
Thank you very much for your changes! I could found only one last minor.

Could to push potential further changes as new separate commits? It would make the review easier for me :)

I really like your pull request! You had to change a lot of code. However, your changes are easily comprehensible.

File: tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/TaggedLogger.kt (hunk 9 above)
Comment (kahgoh) [reply tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/TaggedLogger.kt]: Nope, that constructor isn't needed anymore. I'll remove it 😉

**Comment by kahgoh** [comment]
I've just pushed the change to remove the internal constructor

**Review by pmwmedia — APPROVED** [review]

**Comment by pmwmedia** [comment]
Thank you for your great work! I have merged your pull request with pleasure.
